v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/DialectAdapter.java

```diff
@@ -15,13 +15,100 @@
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceSchemaReference.Kind;
+import com.google.cloud.teleport.v2.source.reader.io.schema.sources.JdbcSchemaReference;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnType;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 
 /**
  * Interface to support various dialects of JDBC databases.
  *
  * <p><b>Note:</b>As a prt of M2 effort, this interface will expose more mehtods than just extending
  * {@link RetriableSchemaDiscovery}.
  */
-public interface DialectAdapter extends RetriableSchemaDiscovery, UniformSplitterDBAdapter {}
+public interface DialectAdapter extends RetriableSchemaDiscovery, UniformSplitterDBAdapter {
+
+  /**
+   * Discover Tables to migrate. This method could be used to auto infer tables to migrate if not
+   * passed via options.
+   *
+   * @param dataSource Provider for JDBC connection.
+   * @param sourceSchemaReference Source database name and (optionally namespace)
+   * @return The list of table names for the given database.
+   * @throws SchemaDiscoveryException - Fatal exception during Schema Discovery.
+   * @throws RetriableSchemaDiscoveryException - Retriable exception during Schema Discovery.
+   *     <p><b>Note:</b>
+   *     <p>The Implementations must log every exception and generate metrics as appropriate.
+   */
+  default ImmutableList<String> discoverTables(
+      DataSource dataSource, SourceSchemaReference sourceSchemaReference)
+      throws SchemaDiscoveryException, RetriableSchemaDiscoveryException {
+    Preconditions.checkArgument(sourceSchemaReference.getKind() == Kind.JDBC);
+    return discoverTables(dataSource.jdbc(), sourceSchemaReference.jdbc());
+  }
+
+  ImmutableList<String> discoverTables(
+      javax.sql.DataSource dataSource, JdbcSchemaReference sourceSchemaReference)
+      throws SchemaDiscoveryException, RetriableSchemaDiscoveryException;
+
+  /**
+   * Discover the schema of tables to migrate.
+   *
+   * @param dataSource Provider for JDBC connection.
+   * @param schemaReference Source database name and (optionally namespace)
+   * @param tables Tables to migrate.
+   * @return The discovered schema.
+   * @throws SchemaDiscoveryException - Fatal exception during Schema Discovery.
+   * @throws RetriableSchemaDiscoveryException - Retriable exception during Schema Discovery.
+   *     <p><b>Note:</b>
+   *     <p>The Implementations must log every exception and generate metrics as appropriate.
+   */
+  default ImmutableMap<String, ImmutableMap<String, SourceColumnType>> discoverTableSchema(
+      DataSource dataSource, SourceSchemaReference schemaReference, ImmutableList<String> tables)
+      throws SchemaDiscoveryException, RetriableSchemaDiscoveryException {
+    Preconditions.checkArgument(schemaReference.getKind() == Kind.JDBC);
```

Comment:
If this dialect adaptor is agnostic of type of source, this check should not be there. Is this temporary ?

Reply:
Dialect adapter is used only for JDBC and within the jdbc package. It implements SchemaDiscovery which is agnostic to source. Hence we ensure that we take in only JDBC source here.

Dialect adapter is implemented by MySQL, PG  (and later Oracle) etc.

Cassandra/Dynamo only need to implement IOWrapper interface.